### PR TITLE
Allow VisualStudioExperimentationService to initialize on a background thread

### LINKED ISSUE #30892: Package load failure when trying to lauch Navigate To after opening a solution
**Version Used**: VS 16.0 P1 build 28231.51.d16.0

**Steps to Reproduce**:

1. Open a solution in VS without any files open in the editor
2. Ctrl + , to launch Navigate To
3. Search for some type

Note: Problem does not repro if you already had some files from the solution open in the editor at the point where Navigate To was launched.

**Expected Behavior**: Should succeed

**Actual Behavior**: Following package load failure error -

image

On debugging the underlying issue appears to be that `AbstractNavigateToSearchService` is trying to instantiate `VisualStudioExperimentationService` from the background thread resulting in an `InvalidOperationException` from the following `Contract.ThrowIfFalse()` call.

**Dump file**: Available at `\\<internalshare>\\<myfolder>\Dumps\RoslynPackageLoad`

```
      Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.Contract.ThrowIfFalse(bool condition, string message) Line 32  C#
Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.ForegroundThreadAffinitizedObject.AssertIsForeground() Line 56   C#
Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Utilities.ForegroundThreadAffinitizedObject.ForegroundThreadAffinitizedObject(Microsoft.CodeAnalysis.Editor.Shared.Utilities.IThreadingContext threadingContext, bool assertIsForeground) Line 34   C#
> Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Experimentation.VisualStudioExperimentationService.VisualStudioExperimentationService(Microsoft.CodeAnalysis.Editor.Shared.Utilities.IThreadingContext threadingContext, Microsoft.VisualStudio.Shell.SVsServiceProvider serviceProvider) Line 27   C#
              [Native to Managed Transition]      
              [Managed to Native Transition]      
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ReflectionHelpers.Instantiate(System.Reflection.MethodBase ctorOrFactoryMethod, object[] arguments) Line 595  C#
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory.RuntimeExportProvider.RuntimePartLifecycleTracker.CreateValue() Line 517 C#
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.Create() Line 1160             C#
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveNext(Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleState nextState) Line 1255          C#
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.MoveToState(Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleState requiredState) Line 1336   C#
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ExportProvider.PartLifecycleTracker.GetValueReadyToExpose() Line 1061             C#
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.ExportProvider.CreateExport.AnonymousMethod__0() Line 461 C#
              mscorlib.dll!System.Lazy<object>.CreateValue() Line 437         C#
              mscorlib.dll!System.Lazy<object>.LazyInitValue() Line 388       C#
              mscorlib.dll!System.Lazy<object>.Value.get() Line 339             C#
Microsoft.VisualStudio.Composition.dll!Microsoft.VisualStudio.Composition.NetFxAdapters.MefV1ExportProvider.UnwrapExport.AnonymousMethod__1() Line 246 C#
System.ComponentModel.Composition.dll!System.ComponentModel.Composition.Primitives.Export.GetExportedValueCore() Line 247 C#
System.ComponentModel.Composition.dll!System.ComponentModel.Composition.Primitives.Export.Value.get() Line 207 C#
System.ComponentModel.Composition.dll!System.ComponentModel.Composition.ExportServices.GetCastedExportedValue<Microsoft.CodeAnalysis.Host.IWorkspaceService>(System.ComponentModel.Composition.Primitives.Export export) Line 135  C#
System.ComponentModel.Composition.dll!System.ComponentModel.Composition.ExportServices.CreateStronglyTypedLazyOfTM.AnonymousMethod__1() Line 88   C#
mscorlib.dll!System.Lazy<Microsoft.CodeAnalysis.Host.IWorkspaceService>.CreateValue() Line 437 C#
mscorlib.dll!System.Lazy<Microsoft.CodeAnalysis.Host.IWorkspaceService>.LazyInitValue() Line 359 C#
mscorlib.dll!System.Lazy<Microsoft.CodeAnalysis.Host.IWorkspaceService>.Value.get() Line 339 C#
Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Host.Mef.MefWorkspaceServices.GetService<Microsoft.CodeAnalysis.Experiments.IExperimentationService>() Line 49   C#
Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.RemoteHostClientExtensions.IsOutOfProcessEnabled(Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.Options.Option<bool> featureOption) Line 129              C#
Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.RemoteHostClientExtensions.TryGetRemoteHostClientAsync(Microsoft.CodeAnalysis.Workspace workspace, Microsoft.CodeAnalysis.Options.Option<bool> featureOption, System.Threading.CancellationToken cancellationToken) Line 135         C#
Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.NavigateTo.AbstractNavigateToSearchService.TryGetRemoteHostClientAsync(Microsoft.CodeAnalysis.Project project, System.Threading.CancellationToken cancellationToken) Line 47           C#
Microsoft.CodeAnalysis.Features.dll!Microsoft.CodeAnalysis.NavigateTo.AbstractNavigateToSearchService.SearchProjectAsync(Microsoft.CodeAnalysis.Project project, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Document> priorityDocuments, string searchPattern, System.Collections.Immutable.IImmutableSet<string> kinds, System.Threading.CancellationToken cancellationToken)        C#
Microsoft.CodeAnalysis.EditorFeatures.Wpf.dll!Microsoft.CodeAnalysis.Editor.Implementation.NavigateTo.NavigateToItemProvider.Searcher.SearchAsyncWorker(Microsoft.CodeAnalysis.Project project, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Document> priorityDocuments) Line 189           C#
Microsoft.CodeAnalysis.EditorFeatures.Wpf.dll!Microsoft.CodeAnalysis.Editor.Implementation.NavigateTo.NavigateToItemProvider.Searcher.SearchAsync(Microsoft.CodeAnalysis.Project project, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Document> priorityDocuments)           C#
Microsoft.CodeAnalysis.EditorFeatures.Wpf.dll!Microsoft.CodeAnalysis.Editor.Implementation.NavigateTo.NavigateToItemProvider.Searcher.SearchProjectsInPriorityOrder.AnonymousMethod__5() Line 146 C#
mscorlib.dll!System.Threading.Tasks.Task<System.Threading.Tasks.Task>.InnerInvoke() Line 680 C#
              mscorlib.dll!System.Threading.Tasks.Task.Execute() Line 2498 C#
              mscorlib.dll!System.Threading.Tasks.Task.ExecutionContextCallback(object obj) Line 2861      C#
mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 954 C#
mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 902 C#
              mscorlib.dll!System.Threading.Tasks.Task.ExecuteWithThreadLocal(ref System.Threading.Tasks.Task currentTaskSlot) Line 2827         C#
              mscorlib.dll!System.Threading.Tasks.Task.ExecuteEntry(bool bPreventDoubleExecution) Line 2756            C#
mscorlib.dll!System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem() Line 2704             C#
              mscorlib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch() Line 820              C#
           mscorlib.dll!System.Threading._ThreadPoolWaitCallback.PerformWaitCallback() Line 1161             C#
```

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #30892 

### Customer scenario

When the first Roslyn document opened in the editor is opened via the Navigate To feature, the Roslyn package fails to load and its IDE features are missing substantial functionality.

### Bugs this fixes

#30892
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/725190

### Workarounds, if any

Open a source document before using Navigate To.

### Risk

Low. In the event other initialization code has bugs (fails to follow required threading rules of the host application), the IDE could hang (deadlock) during the test scenario. However, this is unlikely to result in loss of data because changes cannot be made to most source documents prior to opening the first source document.

### Performance impact

Code paths are not altered outside of the broken scenario.

### Is this a regression from a previous update?

Unknown.

### Root cause analysis

Manual testing scenarios should be updated to cover this case.

### How was the bug found?

Internal report.

### Test documentation updated?

Not yet.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs [hunk 1]
```diff
@@ -20,20 +20,31 @@ internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedO
...
         {
-            try
+            object experimentationServiceOpt = null;
+            MethodInfo isCachedFlightEnabledInfo = null;
+
+            threadingContext.JoinableTaskFactory.Run(async () =>
             {
-                _experimentationServiceOpt = serviceProvider.GetService(typeof(SVsExperimentationService));
-                if (_experimentationServiceOpt != null)
+                await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: dumb question.  how is ths better?  Won't this still block the UI thread while this is happening?

**Comment by CyrusNajmabadi** [comment]
Is there a reason we need to have any blocking here?  can we not just make the API async and allow getting this service through normal JTF means one demand?

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs [hunk 2]
```diff
@@ -20,20 +20,31 @@ internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedO
...
+                catch
+                {
+                }
+            });
+
+            _experimentationServiceOpt = experimentationServiceOpt;
+            _isCachedFlightEnabledInfo = isCachedFlightEnabledInfo;
         }
 
         public bool IsExperimentEnabled(string experimentName)
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: it looks like this is the only entrypoint.  can we just make this async?  i walked back from most of hte callers of this, and it seems like they shouldn't have a problem with this being async.

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs [hunk 3]
```diff
@@ -20,20 +20,31 @@ internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedO
...
+
+                try
                 {
-                    _isCachedFlightEnabledInfo = _experimentationServiceOpt.GetType().GetMethod(
-                        "IsCachedFlightEnabled", BindingFlags.Public | BindingFlags.Instance);
+                    experimentationServiceOpt = serviceProvider.GetService(typeof(SVsExperimentationService));
+                    if (experimentationServiceOpt != null)
+                    {
+                        isCachedFlightEnabledInfo = experimentationServiceOpt.GetType().GetMethod(
+                            "IsCachedFlightEnabled", BindingFlags.Public | BindingFlags.Instance);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: aside: is reflection necessary anymore?  It feels pretty weird that we would need this isntead of just calling the real interface now.

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: Note: if that's unpalatable for v16.  I can understand keeping things sync.  But it seems like we should just make this an async interface for master.

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: > dumb question. how is ths better? Won't this still block the UI thread while this is happening?

In the event this type is constructed on the UI thread, the new code is the same as the old code.

In the event this type is constructed on a background thread, the new code will block the background thread while a UI thread operation completes, rather than crash altogether.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: Let me rephrase :)

Because we are still just blocking the UI thread, this doesn't seem like much of an improvement.  I'd prefer that since this is going ot block the UI thread, we stick with what we have, and we make whatever component is calling into this acquire the service at the appropriate time.

That said, my ideal would simply be that this is just all async. 

--

However, i'm guessing since this is dev16 this is being done to quickly fix an issue.  So i wouldn't block things here.  But it seems like we should really be fixing this up for real outside of dev16 :)

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 3 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: ➡️ I wasn't able to find an appropriate definition of `IVsExperimentationService` to use.

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 1 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: > this doesn't seem like much of an improvement

For the impacted scenario, it's the difference between roslyn working, and roslyn not working at all. 😄 

> we make whatever component is calling into this acquire the service at the appropriate time.

MEF does not guarantee the thread on which objects are initialized. The initializer should be caller-thread-agnostic, either by using JTF or by avoiding code that requires the main thread altogether.

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs [hunk 4]
```diff
@@ -14,43 +17,43 @@ namespace Microsoft.VisualStudio.LanguageServices.Experimentation
     [ExportWorkspaceService(typeof(IExperimentationService), ServiceLayer.Host), Shared]
     internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedObject, IExperimentationService
     {
-        private readonly object _experimentationServiceOpt;
-        private readonly MethodInfo _isCachedFlightEnabledInfo;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+
+        private Optional<Func<string, bool>> _isCachedFlightEnabled;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: could consider an async lazy.

Comment (sharwell) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: ➡️ Since a UI thread dependency is involved, I prefer to avoid the confusion between Roslyn's `AsyncLazy<T>` and the JTF-aware one defined in vs-threading.

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: No worries.  Doing that later is totally ok with me :)

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: interesting.  for my own edification, do we think roslyn's Async lazy could not wrap JTF async code?  Or is it more just being paranoid here?  

> and the JTF-aware one defined in vs-threading.

I'd also be fine with this.  It was more about me seeing that there were non-problematic race conditions in teh code below.  AsyncLazy (roslyn or jtf) just seemed like a clean way to expres intent.  but it's not somethign critical at all.

Comment (sharwell) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: > for my own edification, do we think roslyn's Async lazy could not wrap JTF async code?

Yesterday would have been the perfect day to ask this 🎃

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: :D    

👻  `.Result`  🎃 

> I'm honestly not sure

Good enough reason to avoid for now :D

**Comment by sharwell** [comment]
@CyrusNajmabadi I will submit the asynchronous form as a separate pull request targeting **master**, which we can merge after this simpler pull request moves forward

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs [hunk 5]
```diff
@@ -20,20 +20,31 @@ internal class VisualStudioExperimentationService : ForegroundThreadAffinitizedO
...
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
         public VisualStudioExperimentationService(IThreadingContext threadingContext, SVsServiceProvider serviceProvider)
-            : base(threadingContext, assertIsForeground: true)
+            : base(threadingContext)
         {
-            try
+            object experimentationServiceOpt = null;
+            MethodInfo isCachedFlightEnabledInfo = null;
+
+            threadingContext.JoinableTaskFactory.Run(async () =>
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: This is isn't fire-and-forget, right, so this constructor will block on the background thread if the UI thread is blocked? It would seem if we were initializing this on a background thread, we might just be converting this from a crash to a deadlock if the UI thread wasn't pumping when we hit the crashes?

I could understand this might be a tactical fix and that's a risk we're willing to take, but I'd want that very clearly spelled out if it was.

Comment (sharwell) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: > It would seem if we were initializing this on a background thread, we might just be converting this from a crash to a deadlock if the UI thread wasn't pumping when we hit the crashes?

The UI thread doesn't block JTF operations. A deadlock can only occur if someone is failing to follow the rules of the application.

**Comment by jinujoseph** [comment]
Approved to merge once green

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 5 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: > The UI thread doesn't block JTF operations  A deadlock can only occur if someone is failing to follow the rules of the application.

How do we know Roslyn isn't doing a .Result somewhere? Isn't all the code that's calling this today not following JTF rules?

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 4 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: > interesting. for my own edification, do we think roslyn's Async lazy could not wrap JTF async code? Or is it more just being paranoid here?

@CyrusNajmabadi My expectation is no, simply because we're using TaskCreationSource behind the scenes. A JTF.Run() on the UI thread that implicitly waits on an AsyncLazy won't know what to let through and what won't.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: @jasonmalinowski Thanks!

File: src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs (hunk 5 above)
Comment (sharwell) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: I verified locally that 28307.51.d16.0 reproduces this bug, and that the bug does not occur with this change applied.

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Experimentation/VisualStudioExperimentationService.cs]: Sounds good, thanks for doing some testing.

**Comment by sharwell** [comment]
retest this please
